Add E2E scenarios for multi process support

fractalwrench opened the pull request:
## Goal

Adds E2E scenarios that verify bugsnag supports multi process apps. This builds on the work in #1019.

## Changeset

Added `multi_process.feature`, which contains the following scenarios in an activity & service that run in different processes:
  - Handled JVM error
  - Unhandled JVM error
  - Handled NDK error
  - Unhandled NDK error
  - SharedPreferences migration for user/device info

Tested by running the mazerunner scenario locally.

Cawllec commented on features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp:
```diff
@@ -396,4 +396,28 @@ Java_com_bugsnag_android_mazerunner_scenarios_CXXGetJavaDataScenario_activate(JN
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_mazerunner_scenarios_MultiProcessHandledCXXErrorScenario_activate(JNIEnv *env,
+                                                                                           jobject instance) {
+  bugsnag_notify_env(env, (char *)"Achtung!",
```
Can we change this to something more meaningful, so if we get payloads spilling or duplicating into other tests we can tell where they originated?

Cawllec replied:
```suggestion
  bugsnag_notify_env(env, (char *)"activate",
```

Cawllec commented on features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp:
```diff
@@ -396,4 +396,28 @@ Java_com_bugsnag_android_mazerunner_scenarios_CXXGetJavaDataScenario_activate(JN
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_mazerunner_scenarios_MultiProcessUnhandledCXXErrorScenario_user1(JNIEnv *env,
+                                                                                          jobject instance) {
+  bugsnag_set_user_env(env, (char *)"456", (char *)"joe@example.com", (char *)"Joe");
```
As above
```suggestion
  bugsnag_set_user_env(env, (char *)"1", (char *)"1@test.com", (char *)"MultiProcessUnhandledCXXErrorScenario");
```

Cawllec commented on features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp:
```diff
@@ -396,4 +396,28 @@ Java_com_bugsnag_android_mazerunner_scenarios_CXXGetJavaDataScenario_activate(JN
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_mazerunner_scenarios_MultiProcessUnhandledCXXErrorScenario_user2(JNIEnv *env,
+                                                                                          jobject instance) {
+  bugsnag_set_user_env(env, (char *)"123", (char *)"jane@example.com", (char *)"Jane");
```
As above
```suggestion
  bugsnag_set_user_env(env, (char *)"2", (char *)"2@example.com", (char *)"MultiProcessUnhandledCXXErrorScenario");
```

Cawllec commented on features/fixtures/mazerunner/src/main/java/com/bugsnag/android/mazerunner/scenarios/MultiProcessHandledExceptionScenario.kt:
```diff
@@ -20,17 +20,21 @@ internal class MultiProcessHandledExceptionScenario(
+                    eventMetadata
+                )
+            ) {
+                Bugsnag.setUser("123", "jane@example.com", "Jane")
```
Making sure this is more explicit on which user belongs to which process:
```suggestion
                Bugsnag.setUser("2", "background@example.com", "MultiProcessHandledExceptionScenario")
```

Cawllec commented on features/fixtures/mazerunner/src/main/java/com/bugsnag/android/mazerunner/scenarios/MultiProcessHandledExceptionScenario.kt:
```diff
@@ -20,17 +20,21 @@ internal class MultiProcessHandledExceptionScenario(
+                Bugsnag.notify(generateException())
+            }
+        } else {
+            Bugsnag.setUser("456", "joe@example.com", "Joe")
```
As above
```suggestion
            Bugsnag.setUser("2", "foreground@example.com", "MultiProcessHandledExceptionScenario")
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the event "unhandled" is false
+    And the payload field "events.0.metaData.process.name" equals "com.bugsnag.android.mazerunner"
+    And the payload field "events.0.device.id" is stored as the value "first_device_id"
+    And the payload field "events.0.user.id" equals "123"
```
```suggestion
    And the payload field "events.0.user.id" equals "2"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the payload field "events.0.metaData.process.name" equals "com.bugsnag.android.mazerunner"
+    And the payload field "events.0.device.id" is stored as the value "first_device_id"
+    And the payload field "events.0.user.id" equals "123"
+    And the payload field "events.0.user.name" equals "Jane"
```
```suggestion
    And the payload field "events.0.user.name" equals "MultiProcessHandledExceptionScenario"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the payload field "events.0.device.id" is stored as the value "first_device_id"
+    And the payload field "events.0.user.id" equals "123"
+    And the payload field "events.0.user.name" equals "Jane"
+    And the payload field "events.0.user.email" equals "jane@example.com"
```
```suggestion
    And the payload field "events.0.user.email" equals "background@example.com"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+
+    # device ID is shared between processes
+    And the payload field "events.0.device.id" equals the stored value "first_device_id"
+    And the payload field "events.0.user.id" equals "456"
```
```suggestion
    And the payload field "events.0.user.id" equals "1"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    # device ID is shared between processes
+    And the payload field "events.0.device.id" equals the stored value "first_device_id"
+    And the payload field "events.0.user.id" equals "456"
+    And the payload field "events.0.user.name" equals "Joe"
```
```suggestion
    And the payload field "events.0.user.name" equals "MultiProcessHandledExceptionScenario"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the payload field "events.0.device.id" equals the stored value "first_device_id"
+    And the payload field "events.0.user.id" equals "456"
+    And the payload field "events.0.user.name" equals "Joe"
+    And the payload field "events.0.user.email" equals "joe@example.com"
```
```suggestion
    And the payload field "events.0.user.email" equals "foreground@example.com"
```

Cawllec commented on features/fixtures/mazerunner/src/main/cpp/bugsnags.cpp:
```diff
@@ -396,4 +396,28 @@ Java_com_bugsnag_android_mazerunner_scenarios_CXXGetJavaDataScenario_activate(JN
+Java_com_bugsnag_android_mazerunner_scenarios_MultiProcessHandledCXXErrorScenario_activate(JNIEnv *env,
+                                                                                           jobject instance) {
+  bugsnag_notify_env(env, (char *)"Achtung!",
+                     (char *)"Das ist nicht gut", BSG_SEVERITY_ERR);
```
```suggestion
                     (char *)"MultiProcessHandledCXXErrorScenario", BSG_SEVERITY_ERR);
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the request is valid for the error reporting API version "4.0" for the "Android Bugsnag Notifier" notifier
+    And the request payload contains a completed handled native report
+    And the payload field "events" is an array with 1 elements
+    And the exception "errorClass" equals "Achtung!"
```
```suggestion
    And the exception "errorClass" equals "activate"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the request payload contains a completed handled native report
+    And the payload field "events" is an array with 1 elements
+    And the exception "errorClass" equals "Achtung!"
+    And the exception "message" equals "Das ist nicht gut"
```
```suggestion
    And the exception "message" equals "MultiProcessHandledCXXErrorScenario"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the request is valid for the error reporting API version "4.0" for the "Android Bugsnag Notifier" notifier
+    And the request payload contains a completed handled native report
+    And the payload field "events" is an array with 1 elements
+    And the exception "errorClass" equals "Achtung!"
```
```suggestion
    And the exception "errorClass" equals "activate"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the request payload contains a completed handled native report
+    And the payload field "events" is an array with 1 elements
+    And the exception "errorClass" equals "Achtung!"
+    And the exception "message" equals "Das ist nicht gut"
```
```suggestion
    And the exception "message" equals "MultiProcessHandledCXXErrorScenario"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the event "unhandled" is true
+    And the payload field "events.0.metaData.process.name" equals "com.bugsnag.android.mazerunner"
+    And the payload field "events.0.device.id" is stored as the value "first_device_id"
+    And the payload field "events.0.user.id" equals "123"
```
```suggestion
    And the payload field "events.0.user.id" equals "2"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the payload field "events.0.metaData.process.name" equals "com.bugsnag.android.mazerunner"
+    And the payload field "events.0.device.id" is stored as the value "first_device_id"
+    And the payload field "events.0.user.id" equals "123"
+    And the payload field "events.0.user.name" equals "Jane"
```
```suggestion
    And the payload field "events.0.user.name" equals "MultiProcessUnhandledCXXErrorScenario"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the payload field "events.0.device.id" is stored as the value "first_device_id"
+    And the payload field "events.0.user.id" equals "123"
+    And the payload field "events.0.user.name" equals "Jane"
+    And the payload field "events.0.user.email" equals "jane@example.com"
```
```suggestion
    And the payload field "events.0.user.email" equals "2@example.com"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the event "unhandled" is true
+    And the payload field "events.0.metaData.process.name" equals "com.example.bugsnag.android.mazerunner.multiprocess"
+    And the payload field "events.0.device.id" equals the stored value "first_device_id"
+    And the payload field "events.0.user.id" equals "456"
```
```suggestion
    And the payload field "events.0.user.id" equals "1"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the payload field "events.0.metaData.process.name" equals "com.example.bugsnag.android.mazerunner.multiprocess"
+    And the payload field "events.0.device.id" equals the stored value "first_device_id"
+    And the payload field "events.0.user.id" equals "456"
+    And the payload field "events.0.user.name" equals "Joe"
```
```suggestion
    And the payload field "events.0.user.name" equals "MultiProcessUnhandledCXXErrorScenario"
```

Cawllec commented on features/full_tests/multi_process.feature:
```diff
@@ -0,0 +1,144 @@
+    And the payload field "events.0.device.id" equals the stored value "first_device_id"
+    And the payload field "events.0.user.id" equals "456"
+    And the payload field "events.0.user.name" equals "Joe"
+    And the payload field "events.0.user.email" equals "joe@example.com"
```
```suggestion
    And the payload field "events.0.user.email" equals "1@example.com"
```

Cawllec commented on features/fixtures/mazerunner/src/main/java/com/bugsnag/android/mazerunner/scenarios/SharedPrefMigrationScenario.kt:
```diff
@@ -0,0 +1,69 @@
+        if (isRunningFromBackgroundService()) {
+            prefs.edit()
+                .putString("install.iud", "267160a7-5cf2-42d4-be21-969f1573ecb0")
+                .putString("user.id", "456")
```
We should change these user details to be more overtly meaningful like the others.

Cawllec reviewed (commented):
Looks good.  Most of the following comments are about making the information in the users and error messages more meaningful when test failures occur, especially when multiple users are present in the same test.

One question, is the order of the payloads guaranteed when generating them from two processes?

fractalwrench commented:
>One question, is the order of the payloads guaranteed when generating them from two processes?

PLAT-5559 has been raised to convert each process to deliver to a separate mazerunner endpoint, but this will require some test infrastructure work first. The scenarios have also been written in a way that should mean the request order remains the same.

Addressed all other feedback, rebased, and ran the multiprocess test scenarios locally.
